### PR TITLE
Implemented Linear Controller and added unit test.

### PR DESCRIPTION
**Pull request by MrHohn**
The autoscaler are now supporting one more control pattern: `linear mode`. An example for the ConfigMap params as below:

```
data:
  linear: |-
    { 
      "coresPerReplica": 2,
      "nodesPerReplica": 1,
      "min": 1,
      "max": 100
    }
```

Because serious performance evaluation of `kube-dns` might not be finished soon, I would like to use this simple pattern for the DNS horizontal scaling for now (will tune the specific params). Details for this new pattern are updated in README.

@bowei @thockin @bprashanth 
cc @matchstick

### CODE REVIEW

**Comment by thockin** [comment]
this is net simpler, which is good.  I forget why Girish wanted the ladder mode.  Do we need to keep it?

**Comment by MrHohn** [comment]
I don't think he gave out the reason. But I think it does not harm to keep it.

**Comment by thockin** [comment]
Dead code is a liability - keeping it around means we support it and debug
it and compile it and test it.  If we really think linear is sufficient,
and I don't see why not, we should remove the ladder.  YAGNI.

https://dougseven.com/2014/04/17/knightmare-a-devops-cautionary-tale/

**Comment by MrHohn** [comment]
Your concern totally makes sense, but I vote for keeping the `ladder mode`:
- I won't say `ladder mode` is completely useless. It is, in my opinion, another common and intuitive control pattern.
- `linear mode` might be sufficient for `kube-dns` scaling. But I would like to see other applications also use this autoscaler and they may have different requirements.
- `ladder mode` does something that `linear mode` can not do. It provides users more flexibility and customization. Before we have better substitute for this, I would like to support this pattern.

**Comment by bowei** [comment]
I would agree with keeping ladder as it is a way out for people to implement their own arbitrary scaling function

**Comment by thockin** [comment]
OK.

File: pkg/autoscaler/controller/linearcontroller/linear_controller.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+		}
+		glog.V(4).Infof("Parsed new params: %v\n", params)
+		c.params = params
+		c.version = configMap.Version
+	}
+	return nil
+}
+
+// parseParams Parse the params from JSON string
+func parseParams(data []byte) (params *linearParams, err error) {
```
Comment (bowei) [inline pkg/autoscaler/controller/linearcontroller/linear_controller.go]: you don't reference `params` or `err`, probably should leave them anonymous

File: pkg/autoscaler/controller/linearcontroller/linear_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+		Min:             2,
+		Max:             100,
+	}
+
+	testCases := []struct {
+		controller   *LinearController
+		numResources int
+		expReplicas  int
+	}{
+		{testController, 0, 2},
```
Comment (bowei) [inline pkg/autoscaler/controller/linearcontroller/linear_controller_test.go]: Why is testController the same for all the testcases? can we remove this field from the struct?

File: pkg/autoscaler/controller/linearcontroller/linear_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+		if err != nil && !tc.expError {
+			t.Errorf("unexpected parse failure")
+			spew.Dump(tc)
+			continue
+		}
+		verifyParams(t, params, tc.expParams)
+	}
+}
+
+func TestControllerScaler(t *testing.T) {
```
Comment (bowei) [inline pkg/autoscaler/controller/linearcontroller/linear_controller_test.go]: no testing of max between linear on nodes and linear on pods

File: pkg/autoscaler/controller/linearcontroller/linear_controller.go (hunk 1 above)
Comment (MrHohn) [inline pkg/autoscaler/controller/linearcontroller/linear_controller.go]: Fixed.

File: pkg/autoscaler/controller/linearcontroller/linear_controller_test.go (hunk 3 above)
Comment (MrHohn) [inline pkg/autoscaler/controller/linearcontroller/linear_controller_test.go]: Added one more test.

File: README.md [hunk 4]
```diff
@@ -28,62 +28,103 @@ Usage of cluster-proportional-autoscaler:
...
+  linear: |-
+    { 
+      "coresPerReplica": 2,
+      "nodesPerReplica": 1,
+      "min": 1,
+      "max": 100
+    }
+```
+
+The idea of linear control mode is simple: how many cores or nodes could be taking care by one replica?
```
Comment (bowei) [inline README.md]: I think you should put the equation and then give the example

replicas = max( cores * 1/coresPerReplica , nodes * 1/nodesPerReplica )

File: autoscaler-ladder-example.yaml [hunk 5]
```diff
@@ -77,7 +77,7 @@ spec:
         app: autoscaler
     spec:
       containers:
-        - image: gcr.io/google_containers/cluster-proportional-autoscaler-amd64:0.7.2
+        - image: gcr.io/google_containers/cluster-proportional-autoscaler-amd64:0.8.2
```
Comment (bowei) [inline autoscaler-ladder-example.yaml]: why is the patch number still .2? I would think this will be 0.8.0

File: autoscaler-linear-example.yaml [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    app: autoscaler
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: autoscaler
+    spec:
+      containers:
+        - image: gcr.io/google_containers/cluster-proportional-autoscaler-amd64:0.8.2
```
Comment (bowei) [inline autoscaler-linear-example.yaml]: fix it here as well

File: README.md (hunk 4 above)
Comment (MrHohn) [inline README.md]: Yeah, an equation will explain the idea more clearly. Added.

File: autoscaler-ladder-example.yaml (hunk 5 above)
Comment (MrHohn) [inline autoscaler-ladder-example.yaml]: Fixed.

File: autoscaler-linear-example.yaml (hunk 6 above)
Comment (MrHohn) [inline autoscaler-linear-example.yaml]: Fixed.

**Comment by MrHohn** [comment]
Thanks! Self merging.

**Pull request merged**